torch/lib/c10d/ProcessGroupGloo.cpp

```diff
@@ -953,8 +977,89 @@ class AsyncScatterWork : public ProcessGroupGloo::AsyncWork {
     GENERATE_ALL_TYPES(scalarType, setOutput, opts, outputs[0]);
     gloo::scatter(opts);
   }
+
+  void run() override {
+    scatter(outputs, inputs);
+  }
 };
 
+#ifdef USE_CUDA
+
+class AsyncScatterCUDAWork : public AsyncScatterWork {
+ public:
+  AsyncScatterCUDAWork(
+      const std::shared_ptr<gloo::Context>& context,
+      std::vector<at::Tensor>& outputs,
+      std::vector<std::vector<at::Tensor>>& inputs,
+      int root,
+      uint32_t tag)
+      : AsyncScatterWork(context, outputs, inputs, root, tag) {
+    initializeStreamsEvents(inputs, inputStreams, inputEvents);
+    initializeStreamsEvents(outputs, outputStreams, outputEvents);
```

Comment:
This uses the initializeStreamsEvents function but with inputs/outputs switched around (per the argument names in that function). I think those argument names should be different to avoid confusion. If you look only at the scatter function you would think that the call talks about input goes to the implementation that talks about input, but it goes to the implementation that talks about output.

E.g. they can be called just "tensors" or something similarly opaque. It's not the greatest name but will avoid any possible input/output confusion.

Reply:
resolved in pull request 14576